typegate/src/services/rest_service.ts

```diff
@@ -95,3 +113,120 @@ export async function handleRest(
     }
   }
 }
+
+export function buildOpenAPISpecFrom(baseUrl: string, engine: Engine): string {
+  // https://swagger.io/specification/
+  // https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.0/petstore.json
+  const title = engine.tg.type(0).title;
+  const spec = {
+    openapi: "3.0.3",
+    info: {
+      title: title,
+      license: { name: "MIT" },
+      description: `Rest endpoint for typegraph "${title}"`,
+      version: "1.0.0",
+    },
+    // list server objects
+    servers: [{ url: baseUrl }],
+    // available paths and operations for the API.
+    paths: {} as Record<string, unknown>,
+    // hold various schema for the document
+    components: {},
+  };
+
+  // 502, 400, 403
+  const errorSchema = {
+    type: "object",
+    required: ["message", "extensions"],
+    properties: {
+      message: { type: "integer" },
+      locations: { type: "array", items: { type: "string" } },
+      path: { type: "array", items: { type: "string" } },
+      extensions: {
+        type: "object",
+        properties: {
+          timestamp: { type: "string" },
+        },
+      },
+    },
+  };
+
+  spec.components = {
+    schemas: {
+      Error: errorSchema,
+      GraphQLOutput: {},
+      OpenAPISpec: {},
+    },
+  };
+
+  const genericErrorResponse = (name: string, code: number) => ({
+    description: {
+      400: `Perform ${name}: Bad Request`,
+      403: `Perform ${name}: Forbidden`,
+      500: `Perform ${name}: Service unavailable`,
+    }[code] ?? `Perform ${name}: FAILED`,
+    content: {
+      "application/json": {
+        schema: { $ref: "#/components/schemas/Error" },
+      },
+    },
+  });
+
+  // construct paths
+  for (const method of Object.keys(engine.rest)) {
+    const queries = engine.rest[method];
+    for (const name of Object.keys(queries)) {
+      const [, , getVariables] = queries[name];
+
+      const parameters = getVariables().map((v) => ({
+        name: v,
+        in: "query",
+        required: true,
+        schema: { type: "string" },
+      }));
+
+      spec.paths[`/${title}/rest/${name}`] = {
+        [method.toLowerCase()]: {
+          summary: `Perform ${name}`,
+          operationId: [method, title, name].join("_").toLowerCase(),
+          parameters,
+          responses: {
+            200: {
+              description: `Perform ${name} OK`,
+              content: {
+                "application/json": {
+                  schema: { $ref: "#/components/schemas/GraphQLOutput" },
+                },
+              },
+            },
+            500: genericErrorResponse(name, 500),
+            403: genericErrorResponse(name, 403),
+            400: genericErrorResponse(name, 400),
+          },
+        },
+      };
+    }
+  }
+
+  spec.paths[`/${title}/rest/__rest`] = {
```

Comment:
I believe you can remove that one (as in graphql the introspection does not explain itself)